### PR TITLE
Restructured and organized; address issue 1252

### PR DESCRIPTION
**Pull request by dmatrix**
Signed-off-by: Jules Damji <dmatrix@comcast.net>

## What changes are proposed in this pull request?

* Restructured and organized the table of contents in a cohesive and related subcomponent for MLflow
* Contributions guidelines and governance for MLflow community contributors
* Address #1252 

The new TOC:

Screen Shot 2020-08-03 at 3 46 05 PM

### CODE REVIEW

File: CONTRIBUTING.rst
Comment (harupy) [inline CONTRIBUTING.rst:380]: We shoul rename `traivs` folder because we no longer use it. I'll make a PR.

File: CONTRIBUTING.rst
Comment (harupy) [inline CONTRIBUTING.rst:291]: Should we remove this block because pytest should be installed in the "General prerequisites and dependencies" step?

Comment (dmatrix) [reply CONTRIBUTING.rst:291]: We could since you will need it for Python, in general, so might as well. Will do it.

File: CONTRIBUTING.rst
Comment (dmatrix) [reply CONTRIBUTING.rst:380]: Okay, I'll wait for your PR to make that change since I can't rename it until your PR is merged.

File: CONTRIBUTING.rst
Comment (dmatrix) [reply CONTRIBUTING.rst:291]: done!

**Comment by dmatrix** [comment]
@harupy Thanks for filing the PR for folder change->travis. Do I need to remove all references of mention of `Travis` ->`Dev`? We still use Travis, though (the folder references has been changed)

What about references to `.travis.R` or `.travis.yml`

**Comment by harupy** [comment]
> Do I need to remove all references of mention of Travis ->Dev?

We should because we no longer use Travis at all.

File: CONTRIBUTING.rst
Comment (harupy) [reply CONTRIBUTING.rst:381]: Addressed by #3253

File: CONTRIBUTING.rst
Comment (arjundc-db) [inline CONTRIBUTING.rst:121]: Do you think it would be good to say why we ask them to ask their work?

Comment (dmatrix) [reply CONTRIBUTING.rst:121]: We mention that in the[ Sign your work](https://github.com/mlflow/mlflow/blob/master/CONTRIBUTING.rst#sign-your-work) section: " You need to sign that you wrote the patch or otherwise have the right to pass it on as an open-source patch."

File: CONTRIBUTING.rst
Comment (arjundc-db) [inline CONTRIBUTING.rst:160]: For completeness, can we add an introduction like : 
The MLflow UI is written in JavaScript. `npm` ...

File: CONTRIBUTING.rst
Comment (arjundc-db) [inline CONTRIBUTING.rst:178]: Do you think we need to have separate headings for Launching the Dev UI and Running the JS Dev Server. Can they be merged?

File: CONTRIBUTING.rst
Comment (arjundc-db) [inline CONTRIBUTING.rst:318]: I dont think travis runs the tests anymore. Github actions for the win!
```suggestion
    # Optionally, run large tests as well. Github actions will run large tests on your pull request once
```

File: CONTRIBUTING.rst
Comment (arjundc-db) [inline CONTRIBUTING.rst:374]: Remove reference to travis here and the line above.

File: CONTRIBUTING.rst
Comment (arjundc-db) [inline CONTRIBUTING.rst:447]: Can we add a link for a new user to understand what this means? They may not know what is MLflow project, etc

File: CONTRIBUTING.rst
Comment (arjundc-db) [inline CONTRIBUTING.rst:432]: I dont think we need to nest this heading under Python. Let me know what you think!

File: CONTRIBUTING.rst
Comment (arjundc-db) [inline CONTRIBUTING.rst:295]: Can the section related to writing and running python test come after the Python Client and Python Server? Makes a little more sense to me but open to suggestions!

**Review by arjundc-db — COMMENTED** [review]
This is really awesome @dmatrix , looks and flows so much cleaner now!

File: CONTRIBUTING.rst
Comment (dmatrix) [reply CONTRIBUTING.rst:160]: That sounds reasonable.

File: CONTRIBUTING.rst
Comment (dmatrix) [reply CONTRIBUTING.rst:178]: The original document was written by @smurching, and when we restructured it we placed it as such. I'll leave that up to @smurching if it makes sense to merge them.

File: CONTRIBUTING.rst
Comment (dmatrix) [reply CONTRIBUTING.rst:318]: There places where we will have to change all occurrences of how Travis is replaced with Github's actions.

File: CONTRIBUTING.rst
Comment (dmatrix) [reply CONTRIBUTING.rst:374]: Will delete this line if we don't use it. But there's also reference to `NOT_CRAN=true LINTR_COMMENT_BOT=false Rscript ../.travis.R`. What to do about that one?
Any suggestions?

File: CONTRIBUTING.rst
Comment (dmatrix) [reply CONTRIBUTING.rst:447]: good point! will do

File: CONTRIBUTING.rst
Comment (dmatrix) [reply CONTRIBUTING.rst:432]: Most of these examples are in Python, and @smurching saw it as fitting to put it under Python, at least for now.

File: CONTRIBUTING.rst
Comment (dmatrix) [reply CONTRIBUTING.rst:295]: As they apply to both client and server, we decided to put them before.

File: CONTRIBUTING.rst
Comment (dmatrix) [reply CONTRIBUTING.rst:111]: I suppose you can drop the proposition.

File: CONTRIBUTING.rst
Comment (dmatrix) [reply CONTRIBUTING.rst:151]: Yes, more precise.

File: CONTRIBUTING.rst
Comment (dmatrix) [reply CONTRIBUTING.rst]: done!

File: CONTRIBUTING.rst
Comment (smurching) [inline CONTRIBUTING.rst:111]: IMO the original wording here is actually a bit clearer, is there a reason to change it? With "Developing and testing MLflow" it's a bit less clear that we're testing changes to MLflow, as opposed to using MLflow to develop another project

File: CONTRIBUTING.rst
Comment (smurching) [inline CONTRIBUTING.rst:38]: If we stick with the original wording per https://github.com/mlflow/mlflow/pull/3219/files#r472528940, we should undo this change

File: CONTRIBUTING.rst
Comment (smurching) [inline CONTRIBUTING.rst:117]: Any reason to reword this? The original wording seemed a bit more concise

File: CONTRIBUTING.rst
Comment (dmatrix) [reply CONTRIBUTING.rst:111]: One argument would be while the proposition "to" or "for" makes it specific to MLflow, we are already talking about contributions to MLFlow. The document is about contributing to MLflow. So it may be just redundant. However, the CONTRIBUTIONS.RST was for MLflow and say, Spark MLlib, then having that specificity would make it more clear.

File: CONTRIBUTING.rst
Comment (dmatrix) [reply CONTRIBUTING.rst:117]: I had taught about general (applies to all) and specific to individual components. For example, git applies to all; installing JavaScript dependencies is more specific to that component.

File: CONTRIBUTING.rst
Comment (smurching) [reply CONTRIBUTING.rst:111]: Gotcha, yeah seems fine given that we explicitly state that this page is for "contributing to MLflow" at the top

File: CONTRIBUTING.rst
Comment (smurching) [reply CONTRIBUTING.rst:38]: Given that we're ok with the wording "Developing and testing MLflow" as the header, can we fix this link? I think it should be https://github.com/mlflow/mlflow/blob/master/CONTRIBUTING.rst#developing-and-testing-mlflow, to match https://github.com/mlflow/mlflow/blob/ebacfdaf0095bf816a19f24a64319d6570d81f26/CONTRIBUTING.rst#developing-and-testing-mlflow

File: CONTRIBUTING.rst
Comment (smurching) [reply CONTRIBUTING.rst:117]: Ah got it, sorry makes sense in the context of the surrounding content - maybe "Common prerequisites and dependencies"

File: CONTRIBUTING.rst
Comment (smurching) [inline CONTRIBUTING.rst:276]: Can we replace this sentence ("Other Java functionality (like artifact storage) depends on the Python package, so first install the Python package in a conda environment as described above.") with a link to the "Common prerequisites and dependencies" section, as you've done for Python [here](https://github.com/mlflow/mlflow/pull/3219/files#diff-ea1f65a8eaac0cec7ae151dfdbbbc587R292)? And could we do the same for R?

File: CONTRIBUTING.rst
Comment (smurching) [inline CONTRIBUTING.rst:298]: Can we delete the section "MLflow currently uses ``pytest==3.2.1`` for testing?" Seems unlikely to be immediately useful to readers of this guide & hard to keep up-to-date

File: CONTRIBUTING.rst
Comment (smurching) [inline CONTRIBUTING.rst:307]:
```suggestion
By default, runs are logged under a local temporary directory that's unique to each test and torn down immediately after test execution. To disable this behavior, decorate your test function with ``@pytest.mark.notrackingurimock``
```

File: CONTRIBUTING.rst
Comment (smurching) [inline CONTRIBUTING.rst:335]:
```suggestion
via the ``@pytest.mark.large`` annotation. Dependencies for small and large tests can be added to
```

Sorry to add new suggestions, this just seemed like a good one to fix

File: CONTRIBUTING.rst
Comment (smurching) [inline CONTRIBUTING.rst:443]: Maybe we can remove this reference to PEP8 coding style, as @harupy's linting work now ensures that we lint our examples

File: CONTRIBUTING.rst
Comment (smurching) [reply CONTRIBUTING.rst:440]: Oops, sorry - agree that it makes sense for examples to go into their own top-level section, i.e. "Developing and testing MLflow -> Writing MLflow examples", sorry for the confusion there.

Comment (smurching) [reply CONTRIBUTING.rst:440]: Looks like if a reader follows the "general prereqs" step, they'll actually install pytest & so be able to run the examples tests / develop examples

File: CONTRIBUTING.rst
Comment (smurching) [inline CONTRIBUTING.rst:446]:
```suggestion
To facilitate review, strive for brief examples that reflect real user workflows, document how to run your example, and follow the recommended steps below.
```

File: CONTRIBUTING.rst
Comment (smurching) [inline CONTRIBUTING.rst:450]:
```suggestion
If you are contributing a new model flavor, follow these steps:
```

File: CONTRIBUTING.rst
Comment (smurching) [inline CONTRIBUTING.rst:464]: We don't actually want to encourage people to add new code to the quickstart (editing the existing quickstart is fine, but we should keep it brief) - maybe we should remove this section? Or just have a sentence stating that we welcome changes to the quickstart that make it clearer or simpler etc.

File: CONTRIBUTING.rst
Comment (smurching) [inline CONTRIBUTING.rst:466]:
```suggestion
If you'd like to provide an example of functionality that doesn't fit into the above categories, follow these steps:
```

**Review by smurching — COMMENTED** [review]
Mostly looks good! Just small comments, let's merge once they're addressed - this is a solid improvement over the existing content :) thanks @dmatrix!

File: CONTRIBUTING.rst
Comment (dmatrix) [reply CONTRIBUTING.rst:440]: Excellent! Will move it under "Developing and testing MLflow"

File: CONTRIBUTING.rst
Comment (dmatrix) [reply CONTRIBUTING.rst:117]: Yes, "Common prerequisites and dependencies" makes sense

File: CONTRIBUTING.rst
Comment (dmatrix) [reply CONTRIBUTING.rst:276]: sounds good; will do it.

File: CONTRIBUTING.rst
Comment (dmatrix) [reply CONTRIBUTING.rst:298]: Okay, done!

File: CONTRIBUTING.rst
Comment (dmatrix) [reply CONTRIBUTING.rst:307]: done!

File: CONTRIBUTING.rst
Comment (dmatrix) [reply CONTRIBUTING.rst:335]: No worries, good to fix it. Done!

File: CONTRIBUTING.rst
Comment (dmatrix) [reply CONTRIBUTING.rst:443]: cool, will do!

File: CONTRIBUTING.rst
Comment (dmatrix) [reply CONTRIBUTING.rst:446]: I like that! done!

File: CONTRIBUTING.rst
Comment (dmatrix) [reply CONTRIBUTING.rst:450]: done!

File: CONTRIBUTING.rst
Comment (dmatrix) [reply CONTRIBUTING.rst:466]: done!

**Comment by dmatrix** [comment]
@smurching, I believe I have incorporated all your feedback, along with @harupy's and @arjundc-db 's.  Thanks for all your valuable feedback.

 If I missed anything or you find something that could be remedied, please let me know, and I'll fix it.

**Review by smurching — APPROVED** [review]
LGTM thanks @dmatrix!

**Pull request merged**
